src/lib/gadgets/rot.ts

```diff
@@ -0,0 +1,113 @@
+import { Field } from '../field.js';
+import { Provable } from '../provable.js';
+import { Fp } from '../../bindings/crypto/finite_field.js';
+import * as Gates from '../gates.js';
+
+export { rot, rotate };
+
+const MAX_BITS = 64 as const;
+
+function rot(word: Field, bits: number, direction: 'left' | 'right' = 'left') {
+  // Check that the rotation bits are in range
+  if (bits < 0 || bits > MAX_BITS) {
+    throw Error(`rot: expected bits to be between 0 and 64, got ${bits}`);
+  }
+
+  if (direction !== 'left' && direction !== 'right') {
+    throw Error(
+      `rot: expected direction to be 'left' or 'right', got ${direction}`
+    );
+  }
+
+  if (word.isConstant()) {
+    return new Field(Fp.rot(word.toBigInt(), bits, direction === 'left'));
+  }
+  const [rotated] = rotate(word, bits, direction);
+  return rotated;
+}
+
+function rotate(
+  word: Field,
+  bits: number,
+  direction: 'left' | 'right' = 'left'
+): [Field, Field, Field] {
+  // Check as the prover, that the input word is at most 64 bits.
+  Provable.asProver(() => {
+    if (word.toBigInt() > 2 ** MAX_BITS) {
+      throw Error(
+        `rot: expected word to be at most 64 bits, got ${word.toBigInt()}`
```

Comment:
I think we should also check for this case when the inputs are constants (or before, to catch the constant-case)

Reply:
Is that necessary since we check for it [here](https://github.com/o1-labs/o1js/pull/1182/files#diff-50db730c4f28e956a4c9fdf2c7c3a12a1ed77ea519468ceb8d66428afa678b1dR10)?

I figured we wouldn't expose `rotate` anywhere external, so the only entry point is through `rot`, which does that check.

Reply:
maybe I am mistaken but I don't think `rot` checks for the size of  `word`, it only checks that `!(bits < 0 || bits > MAX_BITS)` and `!(direction !== 'left' && direction !== 'right')` or does it?

Reply:
Oh yeah, you're right. Sorry, I misunderstood!